insights/specs/default.py

```diff
@@ -583,29 +621,46 @@ def package_and_java(broker):
     rpm_V_packages = simple_command("/usr/bin/rpm -V coreutils procps procps-ng shadow-utils passwd sudo", keep_rc=True)
     rsyslog_conf = simple_file("/etc/rsyslog.conf")
     samba = simple_file("/etc/samba/smb.conf")
-    saphostctl_listinstances = simple_command("/usr/sap/hostctrl/exe/saphostctrl -function ListInstances")
+    saphostctrl_listinstances = simple_command("/usr/sap/hostctrl/exe/saphostctrl -function ListInstances")
 
-    @datasource(saphostctl_listinstances, hostname, context=HostContext)
-    def sap_hana_sid(broker):
+    @datasource(saphostctrl_listinstances, hostname)
+    def sap_sid_nr(broker):
         """
-        Command: Get the SID for running "HDB version".
+        Get the SID and Instance Number
 
-        Typical output of saphostctl_listinstances::
+        Typical output of saphostctrl_listinstances::
         # /usr/sap/hostctrl/exe/saphostctrl -function ListInstances
         Inst Info : SR1 - 01 - liuxc-rhel7-hana-ent - 749, patch 418, changelist 1816226
 
+        Returns:
+            (list): List of tuple of SID and Instance Number.
+
         """
-        hana_ins = broker[DefaultSpecs.saphostctl_listinstances].content
+        insts = broker[DefaultSpecs.saphostctrl_listinstances].content
         hn = broker[DefaultSpecs.hostname].content[0].split('.')[0].strip()
         results = set()
-        for ins in hana_ins:
+        for ins in insts:
             ins_splits = ins.split(' - ')
+            # Local Instance
             if ins_splits[2].strip() == hn:
-                results.add(ins_splits[0].split()[-1].lower())
+                # (sid, nr)
+                results.add((ins_splits[0].split()[-1].lower(), ins_splits[1].strip()))
         return list(results)
 
-    sap_hdb_version = foreach_execute(sap_hana_sid, "/usr/bin/sudo -iu %sadm HDB version", keep_rc=True)
+    @datasource(sap_sid_nr)
+    def sap_sid(broker):
+        """
+        Get the SID
+
+        Returns:
+            (list): List of SID.
+
+        """
+        return list(set(sn[0] for sn in broker[DefaultSpecs.sap_sid_nr]))
+
+    sap_hdb_version = foreach_execute(sap_sid, "/usr/bin/sudo -iu %sadm HDB version", keep_rc=True)
     sap_host_profile = simple_file("/usr/sap/hostctrl/exe/host_profile")
+    sapcontrol_getsystemupdatelist = foreach_execute(sap_sid_nr, "/usr/bin/sudo -iu %sadm sapcontrol -nr %s -function GetSystemUpdateList", keep_rc=True)
```

Comment:
This isn't going to work correctly with the simple_file definition in insights_archive. The cat tool will shove all of the outputs into a single file. The definition in insights_archive will read it with simple file, but the parser won't handle it properly.

If you just ran the parser on the live system and let it get results of the foreach_command directly, that's different from what it would get from the simple_file in insights_archive. The single file will cause one parser to be created that's expected to understand how to break the output apart. The direct evaluation will create a separate parser per thing generated by the foreach_execute.

Reply:
The current parser in this PR was for the output of a single command. Yes, I'm sure it needs to be updated once we decided to use the "insights-cat sapcontrol_getsystemupdatelist" command to collect these "foreach_*" specs.

One of my concern is, how to handle the metadata in insights archives and sosreport archives correctly at the same time. E.g. the `ethtool` or `ifcfg` related commands, in sosreport, they are separated in different files (as per `iface`s). If refactor it with `insigths-cat`, the parser has to support the two different format?

Reply:
@csams, please check it again.